zeebe/atomix/cluster/src/test/java/io/atomix/cluster/discovery/DynamicDiscoveryProviderTest.java

```diff
@@ -0,0 +1,221 @@
+    providers.add(provider);
+
+    // when
+    provider.join(null, null).get(10, TimeUnit.SECONDS);
+
+    // then
+    Awaitility.await()
+        .untilAsserted(() -> assertThat(provider.getNodes()).hasSizeGreaterThanOrEqualTo(2));
+  }
+
+  @Test
+  void shouldHandleMultipleResolvedAddress() {
+    // given
+    final DynamicDiscoveryConfig config =
+        new DynamicDiscoveryConfig()
+            .setAddresses(List.of("multi.address.test"))
+            .setRefreshInterval(Duration.ofSeconds(30));
+    final Function<String, List<InetAddress>> mockResolver =
+        ignore -> {
+          try {
+            return List.of(
+                InetAddress.getByName("123.45.66.89"), InetAddress.getByName("123.54.76.98"));
+          } catch (final UnknownHostException e) {
+            throw new RuntimeException(e);
+          }
+        };
+    final DynamicDiscoveryProvider provider = new DynamicDiscoveryProvider(config, mockResolver);
+
+    // when
+    provider.join(null, null);
+    providers.add(provider);
+
+    // then
+    Awaitility.await().until(() -> provider.getNodes().size(), is(2));
+  }
+
+  @Test
+  void shouldCreateNodeWithCorrectId() throws Exception {
+    // given
+    final DynamicDiscoveryConfig config =
+        new DynamicDiscoveryConfig()
+            .setAddresses(List.of("127.0.0.1:26500"))
+            .setRefreshInterval(Duration.ofSeconds(30));
+    final DynamicDiscoveryProvider provider = new DynamicDiscoveryProvider(config);
+    providers.add(provider);
+
+    // when
+    provider.join(null, null).get(10, TimeUnit.SECONDS);
+
+    // then
+    Awaitility.await().until(() -> provider.getNodes().size(), Matchers.equalTo(1));
+    final Set<Node> nodes = provider.getNodes();
+    assertThat(nodes).first().extracting(Node::id).isEqualTo(NodeId.from("127.0.0.1:26500"));
+  }
+
+  @Test
+  void shouldFireJoinEventsOnDiscovery() throws Exception {
+    // given
+    final DynamicDiscoveryConfig config =
+        new DynamicDiscoveryConfig()
+            .setAddresses(List.of("localhost:26500"))
+            .setRefreshInterval(Duration.ofSeconds(30));
+    final DynamicDiscoveryProvider provider = new DynamicDiscoveryProvider(config);
+    providers.add(provider);
+
+    final CountDownLatch latch = new CountDownLatch(1);
+    final List<NodeDiscoveryEvent> events = new ArrayList<>();
+    provider.addListener(
+        event -> {
+          events.add(event);
+          latch.countDown();
+        });
+
+    // when
+    provider.join(null, null).get(10, TimeUnit.SECONDS);
+
+    // then
+    assertThat(latch.await(5, TimeUnit.SECONDS)).isTrue();
+    assertThat(events).isNotEmpty().allMatch(event -> event.type() == Type.JOIN);
+  }
+
+  @Test
+  void shouldNotFailWithUnresolvableAddress() throws Exception {
+    // given
+    final DynamicDiscoveryConfig config =
+        new DynamicDiscoveryConfig()
+            .setAddresses(List.of("unresolvable.invalid.host:26500", "localhost:26500"))
```

Comment:
:thought_balloon: I wonder if this is "slow" when there's no internet connection